Framework: http inflight request dedupe

samouri opened the pull request:
Within the Calypso world, we use `QueryComponents` in order to start network requests for needed assets.  There is nothing stopping many components from firing off duplicate requests for the same asset (at the same time), and/or from a component quickly unmounting/remounting and also causing duplicate requests.

The way we currently solve for this is by tracking inflight requests within the state tree. That means another reducer, reducer tests, selector, and selector tests.  Also, almost all of our query components have a line like this:
```jsx
if ( ! isRequesting )  {
   requestThing( query );
}
```

Now that [http middleware](https://github.com/Automattic/wp-calypso/pull/10327) has landed, we can do better.  We don't need to duplicate this functionality everywhere.  This PR seeks to introduce automatic opt-out inflight request deduping for GET requests.

In order to opt-out of this optimization, you may pass in an extra option to the `http` action creator:

```jsx
http( {
  path: 'chickens',
  options: {
     allowDuplicateInflightRequests: true,
  }
} )
```

blowery commented on client/state/data-layer/wpcom-http/index.js:
```diff
@@ -26,6 +27,7 @@ const fetcherMap = method => get( {
 export const successMeta = data => ( { meta: { dataLayer: { data } } } );
 export const failureMeta = error => ( { meta: { dataLayer: { error } } } );
 export const progressMeta = ( { size: total, loaded } ) => ( { meta: { dataLayer: { progress: { total, loaded } } } } );
+export const requestKey = ( { path, query } ) => JSON.stringify( { type: WPCOM_HTTP_REQUEST, path, query } );
```
this isn't deterministic. key order will vary depending on how the object was created.

blowery replied:
a solution:
```
Object.keys( query )
		.sort() // sort the keys to make the string deterministic. key ordering is not.
		.reduce( ( memo, key ) => memo + `&${ key }=${ query[ key ] }`, '' );
```

blowery replied:
also, if you drop lib/inflight and just use a local `Set`, you can drop the type and just prepend the path to the deterministic query

samouri replied:
> also, if you drop lib/inflight and just use a local Set, you can drop the type and just prepend the path to the deterministic query

do you think that is better? I liked the idea of maintaining inflight separately, but I'm not tied to it

blowery replied:
i don't think we'll need it separately and removing it removes some risk of name clashes

samouri replied:
w.r.t the solution: since this doesn't need to be generic for _any_ object I think we can do something simpler:

```jsx
const requestKey = action => `{
    type: ${ WPCOM_HTTP_REQUEST },
    path: ${ action.path },
    query: ${ action.query }
}`;
```

blowery replied:
where action.query is the deterministically serialized version of the query?

wpcom _may_ support arrays and objects in queries too. need to check.

samouri replied:
> wpcom may support arrays and objects in queries too. need to check.
i'll look into it!

blowery commented on client/state/data-layer/wpcom-http/index.js:
```diff
@@ -37,19 +39,35 @@ const queueRequest = ( { dispatch }, action, next ) => {
+	 * 2. It is a GET request
+	 * 3. There is a currently ongoing request for the same path with the same query
+	 */
+	if ( method === 'GET' && dedupe && inflight.requestInflight( requestKey( action ) ) ) {
```
might want to warn too if dedupe is set to true and the method is not `GET`.

blowery commented on client/state/data-layer/wpcom-http/index.js:
```diff
@@ -37,19 +39,35 @@ const queueRequest = ( { dispatch }, action, next ) => {
+		? inflight.promiseTracker( requestKey( action ), baseRequest )
+		: baseRequest;
+
+	request.then(
```
I feel like @dmsnell had a reason for not using promises here...

dmsnell replied:
the reason here is that if we don't supply a callback then `wpcom.js` will return a `Promise` instead of the actual `request` object and then we won't be able to do things like attach an `onprogress` handler or cancel the request

dmsnell commented on client/state/data-layer/wpcom-http/index.js:
```diff
@@ -37,19 +39,35 @@ const queueRequest = ( { dispatch }, action, next ) => {
 	const method = rawMethod.toUpperCase();
 
-	const request = fetcherMap( method )( ...compact( [
+	/* Eat the request under the following conditions:
```
What does "eat" mean here?

blowery replied:
consume and not pass on.

samouri replied:
eat --> skip

dmsnell commented:
Thanks @samouri!

I'm not going to leave line-level comments here yet because I don't think that we have been able to consider all the right questions yet on what "dedupe" in the data layer should be.

Of the top issues I have right now with this are that it: adds new syntax for what is in truth an optimization, requiring that higher-level things be aware of lower-level concerns (also, `options` in the HTTP request is there for things like this); assumes that in-flight "deduping" should be opt-in.

When I was thinking about this last night I came to a similar mechanism of building a key off of the HTTP request object (or specific properties thereof). However, I wasn't planning on putting it in `lib`. I wanted it to be specific-to and isolated-within the HTTP middleware.

Finally, I think it would be great if we added "dedupe" functionality into its own PR and kept it separate from feature work. Again much of my feelings on this has to do with figuring out what we want to build with this new data system before we start piling up code onto it. Maybe I should add some kind of a "beta" notice.

blowery commented:
> Finally, I think it would be great if we added "dedupe" functionality into its own PR and kept it separate from feature work. Again much of my feelings on this has to do with figuring out what we want to build with this new data system before we start piling up code onto it. Maybe I should add some kind of a "beta" notice.

We need it _for_ the feature work. Usage is the best way to determine what we need.

blowery commented on client/state/data-layer/wpcom-http/actions.js:
```diff
@@ -29,6 +29,7 @@ export const http = ( {
 	onSuccess,
 	onFailure,
 	onProgress,
+	dedupe = false,
```
we can just look for this in `...options`

dmsnell commented:
> We need it for the feature work. Usage is the best way to determine what we need.

understandable @blowery but I don't want to force this onto the data layer before we think it through. we could just stick with `isRequesting` for now and do it the way we've been doing it

blowery commented:
> Of the top issues I have right now with this are that it: adds new syntax for what is in truth an optimization, requiring that higher-level things be aware of lower-level concerns (also, options in the HTTP request is there for things like this); assumes that in-flight "deduping" should be opt-in.

Yeah, I'd be happy with automatically deduping GET requests for the same URL. I'd go the opposite way, and have an option to force the request, even if one is inflight.

blowery commented:
> understandable @blowery but I don't want to force this onto the data layer before we think it through. 

Do you have specific concerns? This is too vague to know when we're ready.

dmsnell commented:
> Do you have specific concerns?

 - I'm concerned about the naming of what we call this: _what_ are we building?
 - should this be opt-in or opt-out?
 - I'm concerned about introducing syntax and function parameters. let's make sure we _really need them_ before we start piling on new things that people have to learn and be aware of

blowery commented:
**What are we building:** 
A way to prevent HTTP GETs for the same URL from being inflight at the same time. This is a pattern we currently reimplement all over calypso, both in flux and redux.

**Opt in or opt out:** 
Opt out, only for GETs

I'm already really sure we really need it. We already have it all over Calypso, reimplemented every time we need it. And we pretty much always need it.

We don't need syntax or function params. We can pass through the opt-out flag in options. Though I can't think of anything off-hand that would require the opt-out.

dmsnell commented:
> I'm already really sure we really need it. We already have it all over Calypso, reimplemented every time we need it. And we pretty much always need it.
>
> We don't need syntax or function params. We can pass through the opt-out flag in options. Though I can't think of anything off-hand that would require the opt-out.

I would say I agree with these. I think it's a safe bet to build this in as such right now in the data-layer. why not create a separate PR though? it can come first. that would likely be a PR which touches maybe two files plus tests and docs and could be reverted in isolation or tracked independently.

again, if it's too much to delay the feature in order to change the architecture, why not just keep the existing pattern and make the architectural change afterwards?

samouri commented:
> Finally, I think it would be great if we added "dedupe" functionality into its own PR and kept it separate from feature work.

I don't mean for this to be feature work as much as I'm trying to couple the dedupe change with an example of it in action.  Isn't it the status quo to give an example with a framework change?

samouri commented:
Changes in latest iterations:
1. removed all `ReaderTeams` code. All thats left is Framework code
2. removed dependency on `lib/inflight`
3. fixed the determinism of `requestKey`
4. opt-in --> opt-out.  new flag is `allowDuplicateInflightRequests`
5. promise --> callback

blowery commented on client/state/data-layer/wpcom-http/index.js:
```diff
@@ -26,8 +26,14 @@ const fetcherMap = method => get( {
 export const successMeta = data => ( { meta: { dataLayer: { data } } } );
 export const failureMeta = error => ( { meta: { dataLayer: { error } } } );
 export const progressMeta = ( { size: total, loaded } ) => ( { meta: { dataLayer: { progress: { total, loaded } } } } );
+export const requestKey = ( { path, query } ) => Object
```
i totally forgot i wrote this but check out `lib/deterministic-stringify`

blowery replied:
off, actually, no, it's broken.

blowery replied:
and i didn't write it! 😄

samouri replied:
are you sure you wrote it? Looks like it was @apeatling!

samouri commented:
punting while we explore other paths for the next bit

dmsnell commented:
@samouri yesterday I merged the HTTP pipeline in #11311. should we close this PR?

samouri commented:
Definitely! Closing 👍
